cli/src/commands/simplify.rs

```diff
@@ -0,0 +1,139 @@
+use std::collections::HashSet;
+
+use itertools::Itertools;
+use jj_lib::commit::Commit;
+use jj_lib::repo::Repo;
+use jj_lib::revset::RevsetExpression;
+use jj_lib::settings::UserSettings;
+
+use crate::cli_util::CommandHelper;
+use crate::cli_util::RevisionArg;
+use crate::cli_util::WorkspaceCommandTransaction;
+use crate::command_error::user_error;
+use crate::command_error::CommandError;
+use crate::ui::Ui;
+
+/// Simplify parent edges for the specified revision(s).
+///
+/// Removes all redundant parent edges going into any of the specified
+/// revisions. By definition, this has no effect on any commit contents and
+/// should not cause any working copy changes or conflicts.
+///
+/// In other words, for all (A, B, C) where A has (B, C) as parents and B has C
+/// as an ancestor, A will be rewritten to have only B as a parent instead of
+/// B+C.
+#[derive(clap::Args, Clone, Debug)]
+pub(crate) struct SimplifyArgs {
+    /// Simplify specified revision(s) together with their trees of descendants
+    /// (can be repeated)
+    #[arg(long, short)]
+    sources: Vec<RevisionArg>,
+    /// Simplify specified revision(s) (can be repeated)
+    #[arg(long, short)]
+    revisions: Vec<RevisionArg>,
+}
+
+pub(crate) fn cmd_simplify(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &SimplifyArgs,
+) -> Result<(), CommandError> {
+    if args.sources.is_empty() && args.revisions.is_empty() {
+        return Err(user_error("no revisions specified"));
+    }
+
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let revs = RevsetExpression::descendants(
+        workspace_command
+            .parse_union_revsets(&args.sources)?
+            .expression(),
+    )
+    .union(
+        workspace_command
+            .parse_union_revsets(&args.revisions)?
+            .expression(),
+    );
+    let commits: Vec<_> = workspace_command
+        .attach_revset_evaluator(revs)
+        .evaluate_to_commits()?
+        .try_collect()?;
+    workspace_command.check_rewritable(commits.iter().map(|c| c.id()))?;
+    let orig_commits = commits.len();
+
+    let mut tx = workspace_command.start_transaction();
+    let mut stats = SimplifyStats::default();
+    for commit in commits {
+        stats.add(simplify_commit(command.settings(), &mut tx, &commit)?);
+    }
```

Comment:
Perhaps, we can use `repo.transform_descendants()` to rewrite commits in topological order. It will save interim predecessor commits.

Maybe we can also build some cache for fast heads computation, but I'm not sure if doing that makes sense. If the history is mostly linear, `index.heads()` might be fast enough.

Reply:
This also lets me re-use the 'simplify_ancestor_merge()' function